packages/belos/tpetra/test/BlockCG/test_bl_cg_hb_multiprec.cpp

```diff
@@ -41,120 +41,126 @@
 // The problem is solver for multiple scalar types, and timings are reported.
 //
-// NOTE: Diagonal preconditioner is used in this case, if diagonals are safe
-//       to invert.
-//
+// NOTE: No preconditioner is used in this case. 
+
 #include "BelosConfigDefs.hpp"
 #include "BelosLinearProblem.hpp"
 #include "BelosTpetraAdapter.hpp"
-#include "BelosBlockCGSolMgr.hpp"
+#include "BelosPseudoBlockCGSolMgr.hpp"
+
+#include <Tpetra_Core.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+#include <Tpetra_DiagPrecond.hpp>
 
 // I/O for Harwell-Boeing files
 #include <Trilinos_Util_iohb.h>
 
 #include <Teuchos_CommandLineProcessor.hpp>
 #include <Teuchos_ScalarTraits.hpp>
-#include <Tpetra_Core.hpp>
-#include <Tpetra_CrsMatrix.hpp>
 #include <Teuchos_TypeNameTraits.hpp>
 
-#include <fstream>
-
 using namespace Teuchos;
 using namespace Belos;
+
 using Tpetra::Operator;
 using Tpetra::CrsMatrix;
+using Tpetra::DiagPrecond;
 using Tpetra::MultiVector;
 using Tpetra::Vector;
 using Tpetra::Map;
+
 using std::endl;
 using std::cout;
 using std::string;
 using std::setw;
 using std::vector;
+
 using Teuchos::tuple;
 
-bool proc_verbose = false, reduce_tol = false, dumpdata = false;
-ParameterList mptestpl;
+bool proc_verbose = false, reduce_tol, precond = true, dumpdata = false;
+RCP<Map<int> > vmap;
+RCP<MultiVector<float,int> > actX;
+ParameterList mptestpl; 
 int rnnzmax;
-int mptestdim, numrhs;
-double *dvals;
+int dim, numrhs; 
+float *fvals;
 int *colptr, *rowind;
 int mptestmypid = 0;
-int mptestnumimages = 1;
 
 ///////////////////////////////////////////////////////////////////////////
-///////////////////////////////////////////////////////////////////////////
-template <class Scalar>
-RCP<LinearProblem<Scalar,Tpetra::MultiVector<Scalar>,Tpetra::Operator<Scalar> > > buildProblem(const RCP<Map<> >& vmap)
+template <class Scalar> 
+RCP<LinearProblem<Scalar,MultiVector<Scalar,int>,Operator<Scalar,int> > > buildProblem()
 {
-  typedef ScalarTraits<Scalar>            SCT;
-  typedef Tpetra::Operator<Scalar>        OP;
-  typedef Tpetra::MultiVector<Scalar>     MV;
-  typedef OperatorTraits<Scalar,MV,OP>    OPT;
-  typedef MultiVecTraits<Scalar,MV>       MVT;
-  using GO = typename MV::global_ordinal_type;
-
-  bool precond = true;
-  RCP<CrsMatrix<Scalar> > A = rcp(new CrsMatrix<Scalar>(vmap,rnnzmax));
-  RCP<CrsMatrix<Scalar> > P = rcp(new CrsMatrix<Scalar>(vmap,1));  // Diagonal preconditioner
+  using SCT = typename ScalarTraits<Scalar>;
+  using MT = typename SCT::magnitudeType;
+  using OP = typename Operator<Scalar,int>;
+  using MV = typename MultiVector<Scalar,int>;
+  using OPT = typename OperatorTraits<Scalar,MV,OP>;
+  using MVT = typename MultiVecTraits<Scalar,MV>;
+
+  RCP<CrsMatrix<Scalar,int> > A = rcp(new CrsMatrix<Scalar,int>(*vmap,rnnzmax));
   if (mptestmypid == 0) {
     // HB format is compressed column. CrsMatrix is compressed row.
-    const double *dptr = dvals;
+    const float *fptr = fvals;
```

Comment:
How about using the magnitude type associated with SCT here? Are we sure float is the right type for these vals?

Reply:
Yes, fixed.